src/solvers/thermal_solver.cpp

```diff
@@ -38,7 +38,9 @@ void ThermalSolver::SetTemperatureBCs(const std::vector<int> &ess_bdr, std::shar
   SetEssentialBCs(ess_bdr, ess_bdr_coef);
 
   // Get the essential dof indicies and project the coefficient onto them
-  temperature.space->GetEssentialTrueDofs(m_ess_bdr, m_ess_tdof_list);
+  for (auto &ess_bc_data : m_ess_bdr) {
+    temperature.space->GetEssentialTrueDofs(ess_bc_data->bc_markers, ess_bc_data->true_dofs);
+  }
```

Comment:
Does `true_dofs` refer to unconstrained degrees of freedom? This becomes a little bit more confusing with multiple entries in `m_ess_bdr`. e.g. a true_dof with respect to `m_ess_bdr[0]` may still be constrained after all.

I'm not sure what to do about it, just thinking aloud.

Reply:
Yeah, I'm considering it an analog to a "node list" that we associate with specific boundary conditions in low order codes. I'm welcome to any ideas to make it clearer! I'm also considering saving the total condensed essential degrees of freedom list saved in the solver object.

Reply:
Disclaimer: This is how I understand MFEM `vdof` vs `tdof` terminology.

I think that `true_dofs` technically means that they are degrees of freedom that are "solved" in Ax=b. In mfem, I believe the `EliminiateRows*` calls just put a 1 on the diagonal, and update the RHS. So you are absolutely correct that any `true_dof` could be constrained.

`virtual_dofs` are those that are degrees of freedom that are "not solved" in Ax=b. They might have been condensed out or etc: ghost nodes, hanging nodes, etc.

Reply:
This is my understanding as well. However, I've been burned before!